src/sql.rs

```diff
@@ -57,7 +58,11 @@ pub struct Sql {
     /// Database file path
     pub(crate) dbfile: PathBuf,
 
-    /// Write transaction mutex.
+    /// Write transactions `Arc`. Used by long-running write transactions to detect a contention
+    /// with other write transactions and if any, try to finish the joob sooner.
+    write_rc: Arc<()>,
```

Comment:
For now i limited it by 1 minute but also respect other write transactions

Reply:
Yes, it will be interrupted, but anyway it runs for 1 minute. But i think, you're right because currently the code is written in a suboptimal way: SQL statements are compiled under the write transactions lock and and as many times as a number of transactions we do in the loop. Btw, i think this should be reworked, statements should be compiled in advance, but it's not a small change.

So, i will drop this `write_rc` and related things. But then we should limit the transaction to 30 seconds, 1 minute of blocking other writers is too much.

As for locks dropage, whether it's done in the right order is not so important really to add tests on that. Even if a context switch happens right upon a mutex unlock (which is unlikely), the transaction is housekeeping will do at least one iteration. Moreover, the order of destructors is documented in Rust and obviously won't change. So i still believe this stuff has sense, but only if we don't do an excessive work like SQL statements compilation on each loop iteration.

Reply:
There's no SQLite API to prepare statements and then execute them on a particular connection. So, generally we have no other options than prepare them under the write transactions lock, otherwise writers will steal connections from readers. But as for housekeeping, we can make an extra connection for it and prepare statements on it _before_ taking write transactions lock. Thus we can release this lock to let other writers make progress while we're running a long transaction and also not recompile necessary statements because of that. But anyway it is likely not needed if we don't run transactions for more than 10s.